### PR TITLE
WIP Add a version of toMap that accepts a mergeFunction

### PR DESCRIPTION
**Pull request by ventuno**
WIP implementation of https://github.com/vavr-io/vavr/issues/2593.

Todo:
- [x] what about `Set`s?
- [x]  what about `toLinkedMap`/`toSortedMap`?
- [x] docs.

### CODE REVIEW

File: src/main/java/io/vavr/collection/Seq.java [hunk 1]
```diff
@@ -1308,4 +1308,21 @@ default boolean startsWith(Iterable<? extends T> that, int offset) {
     default boolean isSequential() {
         return true;
     }
+
+    default <K, V> Map<K, V> toMap(Function<? super T, K> keyMapper, Function<? super T, ? extends V> valueMapper, Function<Seq<? extends V>, ? extends V> mergeFunction) {
+        Objects.requireNonNull(keyMapper, "keyMapper is null");
+        Objects.requireNonNull(valueMapper, "valueMapper is null");
+        Objects.requireNonNull(mergeFunction, "mergeFunction is null");
+
+        return this
```
Comment (ventuno) [inline src/main/java/io/vavr/collection/Seq.java:1317]: An alternative way to do this in Value could be:
```Java
static <T, K, V, E extends Tuple2<? extends K, ? extends V>, R extends Map<K, V>> R toMap(
            Value<T> value, R empty, Function<E, R> ofElement, Function<Iterable<E>, R> ofAll, Function<? super T, ? extends E> f, BiFunction<V,  V, V> mergeFunction) {
        if (value.isEmpty()) {
            return empty;
        } else if (value.isSingleValued()) {
            return ofElement.apply(f.apply(value.get()));
        } else {
            return Iterator.ofAll(value)
                    .map(f)
                    .groupBy(Tuple2::_1)
                    .mapValues(s -> s.map(Tuple2::_2).reduce(mergeFunction));
        }
    }
```
but makes the compiler a bit unhappy.

File: src/main/java/io/vavr/collection/Traversable.java [hunk 2]
```diff
@@ -1704,6 +1704,15 @@ default Number sum() {
      */
     <U> Traversable<U> zipWithIndex(BiFunction<? super T, ? super Integer, ? extends U> mapper);
 
+
+    default <K, V> Map<K, V> toMap(
+            Function<? super T, K> keyMapper,
+            Function<? super T, V> valueMapper,
+            BiFunction<? super V, ? super V, ? extends V> merge) {
+        return this
```
Comment (danieldietrich) [inline src/main/java/io/vavr/collection/Traversable.java:1712]: The functions `groupBy`, `mapValues` and `reduce` already check that the arguments are non-null but I would add explicit checks like you did before:

```java
Objects.requireNonNull(keyMapper, "keyMapper is null");
Objects.requireNonNull(valueMapper, "valueMapper is null");
Objects.requireNonNull(merge, "merge is null");
```

That way error messages contain the correct parameter name.

Please also add the following Javadoc:

```java
    /**
     * Converts this collection to a {@link Map}.
     *
     * @param keyMapper   A function that maps an element to a key
     * @param valueMapper A function that maps an element to a value
     * @param merge A function that merges values that are associated with the same key
     * @param <K>         The key type
     * @param <V>         The value type
     * @return A new {@link HashMap}.
     */
```

Comment (ventuno) [reply src/main/java/io/vavr/collection/Traversable.java:1724]: Done. Thanks!

File: src/main/java/io/vavr/collection/Traversable.java [hunk 3]
```diff
@@ -1704,6 +1704,45 @@ default Number sum() {
...
+     * Converts this collection to a {@link Map}.
+     *
+     * @param f A function that maps an element to a key/value pair represented by Tuple2
+     * @param merge A function that merges values that are associated with the same key
+     * @param <K>         The key type
+     * @param <V>         The value type
+     * @return A new {@link HashMap}.
+     */
+    default <K, V, E extends Function<? super T, ? extends Tuple2<? extends K, ? extends V>>> Map<K, V> toMap(
+            E f,
```
Comment (ventuno) [inline src/main/java/io/vavr/collection/Traversable.java:1717]: This helps to disambiguate between this method and [`Value.toMap(Function, Function)`](https://www.javadoc.io/doc/io.vavr/vavr/0.9.2/io/vavr/Value.html#toMap-java.util.function.Function-java.util.function.Function-).

```
src/main/java/io/vavr/collection/Traversable.java:1716: warning: [overloads] <K#1,V#1>toMap(Function<? super T#1,? extends Tuple2<? extends K#1,? extends V#1>>,BiFunction<? super V#1,? super V#1,? extends V#1>) in Traversable is potentially ambiguous with <K#2,V#2>toMap(Function<? super T#2,? extends K#2>,Function<? super T#2,? extends V#2>) in Value
    default <K, V> Map<K, V> toMap(
                             ^
  where K#1,V#1,T#1,K#2,V#2,T#2 are type-variables:
    K#1 extends Object declared in method <K#1,V#1>toMap(Function<? super T#1,? extends Tuple2<? extends K#1,? extends V#1>>,BiFunction<? super V#1,? super V#1,? extends V#1>)
    V#1 extends Object declared in method <K#1,V#1>toMap(Function<? super T#1,? extends Tuple2<? extends K#1,? extends V#1>>,BiFunction<? super V#1,? super V#1,? extends V#1>)
    T#1 extends Object declared in interface Traversable
    K#2 extends Object declared in method <K#2,V#2>toMap(Function<? super T#2,? extends K#2>,Function<? super T#2,? extends V#2>)
    V#2 extends Object declared in method <K#2,V#2>toMap(Function<? super T#2,? extends K#2>,Function<? super T#2,? extends V#2>)
    T#2 extends Object declared in interface Value
error: warnings found and -Werror specified
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
1 error
1 warning
```
The only alternatives I could find are:
1) change the method name to something like `toMapWithMerge` (but kind of defeats the purpose of aligning this API with Java's);
2) use `SuppressWarnings("overloads")` instead.

Comment (danieldietrich) [reply src/main/java/io/vavr/collection/Traversable.java:1717]: My experience is that similar method signatures will cause problems when used in conjunction with lambdas. Please delete this method for now. We will keep only the second method (below).

File: src/main/java/io/vavr/collection/Traversable.java [hunk 4]
```diff
@@ -1704,6 +1704,45 @@ default Number sum() {
...
+     * @return A new {@link HashMap}.
+     */
+    default <K, V> Map<K, V> toMap(
+            Function<? super T, ? extends K> keyMapper,
+            Function<? super T, ? extends V> valueMapper,
+            BiFunction<? super V, ? super V, ? extends V> merge) {
+        Objects.requireNonNull(keyMapper, "keyMapper is null");
+        Objects.requireNonNull(valueMapper, "valueMapper is null");
+        Objects.requireNonNull(merge, "merge is null");
+        return toMap(t -> Tuple.of(keyMapper.apply(t), valueMapper.apply(t)), merge);
```
Comment (danieldietrich) [inline src/main/java/io/vavr/collection/Traversable.java:1744]:
```java
return map(f)
        .groupBy(keyMapper)
        .mapValues(s -> s.map(valueMapper).reduce(merge));
```

File: src/test/java/io/vavr/collection/CollectionsTest.java [hunk 5]
```diff
@@ -93,6 +96,41 @@ public void shouldNotBeEqualMapAndSet() throws Exception {
...
+
+        Map<String, Integer> m2 = input.toMap(
+                Function.identity(),
+                Integer::sum);
+
+        assertThat(m2.containsKey("a")).isTrue();
+        assertThat(m2.get("a").get()).isEqualTo(60);
+        assertThat(m2.containsKey("b")).isTrue();
+        assertThat(m2.get("b").get()).isEqualTo(2);
+    }
```
Comment (danieldietrich) [inline src/test/java/io/vavr/collection/CollectionsTest.java:132]: Please keep only the first test `m`.

`m2` is obsolete (see above) and `m1` does not bring anything new to the table.

File: src/main/java/io/vavr/collection/Traversable.java (hunk 4 above)
Comment (ventuno) [reply src/main/java/io/vavr/collection/Traversable.java:1744]: Done, but I still need the type witnesses to get this to compile.

File: src/test/java/io/vavr/collection/CollectionsTest.java (hunk 5 above)
Comment (ventuno) [reply src/test/java/io/vavr/collection/CollectionsTest.java:132]: Makes sense, done.

**Comment by ventuno** [comment]
Thanks @danieldietrich. Please take another look when you get a chance :-).

**Comment by danieldietrich** [comment]
Many thanks, that looks fine :)

**Comment by ventuno** [comment]
Thanks @danieldietrich, any chance you'd accept a version of `toLinkedMap`/`toSortedMap` with the `merge` function in a separate PR?

**Comment by danieldietrich** [comment]
@ventuno Please go ahead!

I think there are two challenges:

* the `toLinkedMap` conversion function must maintain key order in iteration order of the original collection, especially when merging collisions. That means that the position of the key is determined when it is put into the Map for the first time. On subsequent merges, the position of the key does not change within the LinkedHashMap. We need a unit test to ensure that.

Screenshot 2020-05-19 at 18 31 45

Please note that we return a `Map`here:

```java
default <K, V> Map<K, V> toLinkedMap(
             Function<? super T, ? extends K> keyMapper,
             Function<? super T, ? extends V> valueMapper,
             BiFunction<? super V, ? super V, ? extends V> merge) {
    ...
}

* there are two versions of the `toSortedMap` conversion:

```java
default <K, V> SortedMap<K, V> toSortedMap(
             Function<? super T, ? extends K> keyMapper,
             Function<? super T, ? extends V> valueMapper,
             BiFunction<? super V, ? super V, ? extends V> merge) {
    return toSortedMap(Comparators.naturalComparator(), keyMapper, valueMapper, merge);
}

default <K, V> SortedMap<K, V> toSortedMap(
             Comparator<? super K> comparator,
             Function<? super T, ? extends K> keyMapper,
             Function<? super T, ? extends V> valueMapper,
             BiFunction<? super V, ? super V, ? extends V> merge) {
    ...
}
```

Thank you for your help!

**Comment by ventuno** [comment]
Thanks @danieldietrich please take a look at https://github.com/vavr-io/vavr/pull/2598 which implements `toSortedMap`.

**Pull request merged**